### PR TITLE
feat: add Kimi token usage observability

### PR DESCRIPTION
**Pull request by nikhilachale**
## What

Adds Kimi token-usage observability using the shared token-source foundation.

- discovers Kimi main and child-agent `wire.jsonl` sources
- parses `usage.record` events into the canonical token model
- persists and aggregates Kimi usage through the shared collector and SQLite store
- adds parser, discovery, migration, and storage coverage

## Why this PR was recreated

The earlier provider PRs were stacked on one another instead of being based cleanly on `main`. That made them difficult to review and prevented them from merging independently into `main`.

This is the clean replacement for the earlier Kimi work. Pi and Qwen were also recreated as separate PRs so each provider can be reviewed, tested, merged, or reverted independently.

## Related PRs and merge order

1. **Kimi:** #4311 (this PR)
2. **Pi:** #4326
3. **Qwen:** #4327

All three target `main`; the order above is recommended for the provider-usage migrations.

## Implementation

Kimi wire usage is normalized through the existing shared token-source contracts. Cache reads remain cached input; direct input plus cache creation form derived uncached input; all input buckets form total input.

The storage migration preserves existing usage rows, IDs, indexes, views, and change-data-capture triggers while widening the supported harness and source-kind constraints.

## Testing

- `go test ./internal/observe/usage ./internal/service/usage ./internal/storage/sqlite ./internal/storage/sqlite/store ./internal/daemon -count=1`
- `go vet ./...`
- focused golangci-lint over all changed packages: `0 issues`
- `git diff --check`

The full backend run only encountered the repository's pre-existing timing-sensitive fake-agent lifecycle test; the same failure was reproduced on unchanged `main`. All directly affected packages pass.

## Checklist

- [x] Based on `main`
- [x] One provider-focused change
- [x] Tests added for the changed behavior
- [x] Relevant local checks pass

### CODE REVIEW

File: backend/internal/storage/sqlite/migrate_burned_versions_test.go [hunk 1]
```diff
@@ -110,6 +110,7 @@ var shippedMigrations = map[int64]string{
 	104: "0104_agent_inventory_cache.sql",
 	105: "0105_default_session_mode_chat.sql",
 	106: "0106_pr_comment_review_id.sql",
+	107: "0107_allow_kimi_usage.sql",
```
Comment (illegalcall) [inline backend/internal/storage/sqlite/migrate_burned_versions_test.go]: [P1] Renumber this migration before merge

`main` already ships migration version 107 as `0107_recovered_conversation_turns.sql`. Reusing 107 here creates a ledger merge conflict and can cause Goose to treat this distinct schema change as already applied on existing installations. Please rebase, assign the next unused migration number, and update this ledger entry.

File: backend/internal/observe/usage/parser_kimi.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+			recordMalformed(result)
+			continue
+		}
+		identity := firstNonEmpty(native.ID, kimiRecordTimeIdentity(native.Time), strconv.FormatInt(record.Offset, 10))
+		event := domain.ModelUsageEvent{
+			ProviderID:      domain.UsageProviderAnthropic,
+			ModelID:         model,
+			Tokens:          tokens,
+			ProviderDetails: domain.UsageProviderDetails{Anthropic: &details},
+			CreatedAt:       parseUsageTimestamp(kimiRecordTimeIdentity(native.Time)),
```
Comment (illegalcall) [inline backend/internal/observe/usage/parser_kimi.go]: [P2] Parse numeric epoch-millisecond timestamps

`kimiRecordTimeIdentity` accepts numeric `time` values, but this passes their decimal representation to `parseUsageTimestamp`, which only accepts RFC3339. Numeric records therefore get a zero `CreatedAt` and are persisted with a NULL timestamp. Please convert numeric values from Unix milliseconds and add a numeric fixture test while retaining RFC3339 support if needed.

File: backend/internal/service/usage/collector.go [hunk 3]
```diff
@@ -1343,6 +1377,48 @@ func (c *Collector) registerDiscoveredClaudeSubagents(
...
 
+func (c *Collector) registerDiscoveredKimiAgents(
+	ctx context.Context,
+	binding domain.UsageBindingRecord,
+	mainPath string,
+	now time.Time,
+	reactivateExisting bool,
+) error {
+	sessionDir := filepath.Dir(filepath.Dir(filepath.Dir(mainPath)))
+	paths, err := filepath.Glob(filepath.Join(sessionDir, "agents", "*", "wire.jsonl"))
```
Comment (illegalcall) [inline backend/internal/service/usage/collector.go:1423]: [P2] Keep Kimi child-agent discovery active

This scan runs only when the main source is initially registered or explicitly reconciled. `HasPendingUsageDiscovery` and `ListUsageDiscoveryBindings` still restrict discovery to Claude/Codex, so a child `wire.jsonl` created after this scan can be missed unless another lifecycle hook happens. Please include Kimi in bounded reconciliation, or watch the agents directory, and test a child file created after initial registration.

File: backend/internal/service/usage/collector.go [hunk 4]
```diff
@@ -1654,18 +1730,67 @@ func validateSourceAttribution(
...
+		subagentID,
+		resolved,
+		expectedParentID,
+	); err != nil {
+		return err
+	}
+	if kind != domain.UsageSourceKimiWire {
+		return nil
+	}
+	expectedRoot, err := filepath.EvalSymlinks(filepath.Join(c.roots.KimiHome, "sessions"))
```
Comment (illegalcall) [inline backend/internal/service/usage/collector.go:1809]: [P2] Add rejection tests for this trust boundary

This containment check protects the provider-owned artifact boundary, but the new tests cover only a valid path. Please add cases for an absolute path outside the Kimi sessions root, a symlink escape, a deleted index record, and malformed index data so a future regression cannot ingest an arbitrary JSONL file.

File: backend/internal/service/usage/collector.go [hunk 5]
```diff
@@ -332,8 +338,11 @@ func (c *Collector) RecordHook(ctx context.Context, sessionID domain.SessionID,
 
 	if mainArtifact != nil {
 		kind := domain.UsageSourceClaudeMain
-		if session.Harness == domain.HarnessCodex {
+		switch session.Harness {
+		case domain.HarnessCodex:
 			kind = domain.UsageSourceCodexRollout
+		case domain.HarnessKimi:
```
Comment (illegalcall) [inline backend/internal/service/usage/collector.go]: [P3] Centralize harness-to-source-kind mapping

This mapping is repeated in the hook, backfill, and reconcile paths. A future provider can be added to one switch and silently default to Claude in another. Please extract one `sourceKindForHarness` helper and cover the supported harnesses once.

**Review by illegalcall — CHANGES_REQUESTED** [review]
Requesting changes based on the inline findings below. The targeted tests pass, but the migration conflict and usage-discovery/timestamp gaps need to be resolved before merge.

File: backend/internal/storage/sqlite/migrate_burned_versions_test.go (hunk 1 above)
Comment (nikhilachale) [reply backend/internal/storage/sqlite/migrate_burned_versions_test.go]: Fixed. Rebased #4311 onto current main, renumbered the Kimi migration to 0108_allow_kimi_usage.sql, and updated the append-only migration ledger.

File: backend/internal/service/usage/collector.go (hunk 3 above)
Comment (nikhilachale) [reply backend/internal/service/usage/collector.go:1423]: Fixed. Active Kimi bindings now remain in bounded discovery reconciliation, and HasPendingUsageDiscovery schedules the retry. Added a regression test that creates a child wire.jsonl after initial registration and discovers it during reconciliation.

File: backend/internal/service/usage/collector.go (hunk 4 above)
Comment (nikhilachale) [reply backend/internal/service/usage/collector.go:1809]: Covered. Added rejection tests for an absolute path outside the Kimi sessions root, a symlink escape, a deleted latest index record, and malformed index data.

File: backend/internal/service/usage/collector.go (hunk 5 above)
Comment (nikhilachale) [reply backend/internal/service/usage/collector.go]: Fixed. Extracted sourceKindForHarness and replaced the repeated hook, backfill, and reconciliation mappings. Added one table test covering supported and unsupported harnesses.

File: backend/internal/observe/usage/parser_kimi.go (hunk 2 above)
Comment (nikhilachale) [reply backend/internal/observe/usage/parser_kimi.go]: Fixed. Kimi timestamps now parse numeric Unix epoch milliseconds while retaining RFC3339 support. Added fixtures covering both formats.

File: backend/internal/storage/sqlite/migrations/0117_allow_kimi_usage.sql [hunk 6]
```diff
@@ -0,0 +1,202 @@
+-- +goose NO TRANSACTION
```
Comment (illegalcall) [inline backend/internal/storage/sqlite/migrations/0117_allow_kimi_usage.sql:1]: **[P1] Rebase this feature onto current `main` and allocate a fresh migration version.** Current `main` already owns migrations 0108 through 0116, so this file collides with an existing migration and the branch also conflicts in the burned-version ledger. This needs more than a numeric rename: current `main` moved usage storage to `usage_measurement_kind` and `provider_usage_json` in migration 0115, so please port the Kimi changes onto that schema and use the next available post-main migration.

File: backend/internal/observe/usage/parser_kimi.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+		return time.Time{}
+	}
+	milliseconds, err := number.Int64()
+	if err != nil {
+		return time.Time{}
+	}
+	return time.UnixMilli(milliseconds).UTC()
+}
+
+func kimiRecordTimeIdentity(raw json.RawMessage) string {
```
Comment (illegalcall) [inline backend/internal/observe/usage/parser_kimi.go]: **[P3] Decode this timestamp once and reuse the result for both `CreatedAt` and source-event identity.** `parseKimiRecordTimestamp` and `kimiRecordTimeIdentity` independently interpret the same raw string/number value. That duplicates the format rules and allows the event timestamp and identity to drift if one decoder changes later. A small decoded value carrying both the canonical identity and `time.Time` would keep them consistent.

**Review by illegalcall — CHANGES_REQUESTED** [review]
Requesting changes because this head is not compatible with current `main`. The earlier epoch-millisecond and active child-session discovery fixes look good, and the exact-head usage tests pass. The remaining blockers are called out inline.

File: backend/internal/storage/sqlite/migrations/0117_allow_kimi_usage.sql (hunk 6 above)
Comment (nikhilachale) [reply backend/internal/storage/sqlite/migrations/0117_allow_kimi_usage.sql:1]: Fixed in e1177b506. Rebased onto current main, replaced the colliding migration with 0117_allow_kimi_usage.sql, and ported the rebuild to preserve the post-0115/0116 schema, including provider_hint, usage_measurement_kind, provider_usage_json, and billing_provider_source. Added a direct 0116-to-0117 migration test that preserves existing usage facts and accepts Kimi bindings/sources.

File: backend/internal/observe/usage/parser_kimi.go (hunk 7 above)
Comment (nikhilachale) [reply backend/internal/observe/usage/parser_kimi.go]: Fixed in 5439b00e7. decodeKimiRecordTime now returns the canonical identity and time.Time together; parseKimi calls it once and reuses both values for SourceEventKey input and CreatedAt. Tests cover RFC3339 and epoch-millisecond inputs.

**Review by illegalcall — APPROVED** [review]
